pylint/extensions/dict_init_mutate.py

```diff
@@ -0,0 +1,118 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Check for use of dictionary mutation after initialization."""
+
+from typing import TYPE_CHECKING, List
+
+from astroid import nodes
+
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import only_required_for_messages
+from pylint.interfaces import HIGH
+from pylint.lint.pylinter import PyLinter
+
+if TYPE_CHECKING:
+    from pylint.lint.pylinter import PyLinter
+
+
+def stringify_dict_items(node: nodes.NodeNG) -> str:
+    try:
+        key = node.targets[0].slice.value
+        value = node.value.value
+    except AttributeError:
+        # More complex cases will be handled later on
+        return ""
+
+    if isinstance(value, str):
+        value = f"'{value}'"
+    return f"'{key}': {value}"
+
+
+class DictInitMutateChecker(BaseChecker):
+    name = "dict-init-mutate"
+    msgs = {
+        "C3301": (
+            "Declare all known key/values when initializing the dictionary with '%s'",
+            "dict-init-mutate",
+            "Dictionaries can be initialized with a single statement "
+            "using dictionary literal syntax.",
+        )
+    }
+
+    @only_required_for_messages("dict-init-mutate")
+    def visit_assign(self, node: nodes.Assign) -> None:
+        """
+        Detect dictionary mutation immediately after initialization.
+
+        At this time, detecting nested mutation is not supported.
+        """
+        if not isinstance(node.value, nodes.Dict):
+            return
+
+        dict_name = node.targets[0]
+        if len(node.targets) != 1 or not isinstance(dict_name, nodes.AssignName):
+            return
+
+        first_sibling = node.next_sibling()
+        if (
+            not first_sibling
+            or not isinstance(first_sibling, nodes.Assign)
+            or len(first_sibling.targets) != 1
+        ):
+            return
+
+        sibling_target = first_sibling.targets[0]
+        if not isinstance(sibling_target, nodes.Subscript):
+            return
+
+        sibling_name = sibling_target.value
+        if not isinstance(sibling_name, nodes.Name):
+            return
+
+        if sibling_name.name == dict_name.name:
+            siblings = self._get_dict_mutating_siblings(first_sibling)
+            suggestion = self._make_dict_suggestion(siblings, dict_name.name)
+            self.add_message(
+                "dict-init-mutate", node=node, args=(suggestion,), confidence=HIGH
+            )
+
+    def _get_dict_mutating_siblings(
+        self, node: nodes.Assign, all_siblings: List[nodes.Assign] | None = None
```

Comment:
Can't use optional-bar syntax in older Python versions :cry:

Reply:
We have python > 3.7.2 we can do it as long as we use ``from __future__ import annotations``

Reply:
AHA that's the magic line!!! I couldn't figure it out, thanks Pierre!